File: mobile/lib/models/quiz.dart
class Choice {
  final int id;
  final String text;

  Choice({required this.id, required this.text});

  factory Choice.fromJson(Map<String, dynamic> json) =>
      Choice(id: json['id'] as int, text: json['text'] as String);
}

class Question {
  final int id;
  final String text;
  final int orderIndex;
  final List<Choice> choices;

  Question({
    required this.id,
    required this.text,
    required this.orderIndex,
    required this.choices,
  });

  factory Question.fromJson(Map<String, dynamic> json) => Question(
    id: json['id'] as int,
    text: json['text'] as String,
    orderIndex: json['order_index'] as int,
    choices: (json['choices'] as List)
        .map((e) => Choice.fromJson(e as Map<String, dynamic>))
        .toList(),
  );
}

class Quiz {
  final int id;
  final int lessonId;
  final String title;
  final int passingScore;
  final List<Question> questions;

  Quiz({
    required this.id,
    required this.lessonId,
    required this.title,
    required this.passingScore,
    required this.questions,
  });

  factory Quiz.fromJson(Map<String, dynamic> json) => Quiz(
    id: json['id'] as int,
    lessonId: json['lesson_id'] as int,
    title: json['title'] as String,
    passingScore: json['passing_score'] as int,
    questions: (json['questions'] as List)
        .map((e) => Question.fromJson(e as Map<String, dynamic>))
        .toList(),
  );
}

class AnswerResult {
  final int questionId;
  final int? selectedChoiceId;
  final int correctChoiceId;
  final bool isCorrect;

  AnswerResult({
    required this.questionId,
    required this.selectedChoiceId,
    required this.correctChoiceId,
    required this.isCorrect,
  });

  factory AnswerResult.fromJson(Map<String, dynamic> json) => AnswerResult(
    questionId: json['question_id'] as int,
    selectedChoiceId: json['selected_choice_id'] as int?,
    correctChoiceId: json['correct_choice_id'] as int,
    isCorrect: json['is_correct'] as bool,
  );
}

class QuizResult {
  final int attemptId;
  final int score;
  final int total;
  final double percentage;
  final bool passed;
  final DateTime submittedAt;
  final List<AnswerResult> answers;

  QuizResult({
    required this.attemptId,
    required this.score,
    required this.total,
    required this.percentage,
    required this.passed,
    required this.submittedAt,
    required this.answers,
  });

  factory QuizResult.fromJson(Map<String, dynamic> json) => QuizResult(
    attemptId: json['attempt_id'] as int,
    score: json['score'] as int,
    total: json['total'] as int,
    percentage: (json['percentage'] as num).toDouble(),
    passed: json['passed'] as bool,
    submittedAt: DateTime.parse(json['submitted_at'] as String),
    answers: (json['answers'] as List)
        .map((e) => AnswerResult.fromJson(e as Map<String, dynamic>))
        .toList(),
  );
}

class QuizAttempt {
  final int id;
  final int quizId;
  final int score;
  final int total;
  final bool passed;
  final DateTime submittedAt;

  QuizAttempt({
    required this.id,
    required this.quizId,
    required this.score,
    required this.total,
    required this.passed,
    required this.submittedAt,
  });

  factory QuizAttempt.fromJson(Map<String, dynamic> json) => QuizAttempt(
    id: json['id'] as int,
    quizId: json['quiz_id'] as int,
    score: json['score'] as int,
    total: json['total'] as int,
    passed: json['passed'] as bool,
    submittedAt: DateTime.parse(json['submitted_at'] as String),
  );
}

File: backend/app/schemas/quiz.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, field_validator


class ChoiceCreate(BaseModel):
    text: str
    is_correct: bool = False


class QuestionCreate(BaseModel):
    text: str
    order_index: int = 0
    choices: list[ChoiceCreate]

    @field_validator("choices")
    @classmethod
    def must_have_one_correct_choice(cls, choices: list[ChoiceCreate]) -> list[ChoiceCreate]:
        if len(choices) < 2:
            raise ValueError("Each question needs at least two choices")
        if sum(1 for c in choices if c.is_correct) != 1:
            raise ValueError("Each question must have exactly one correct choice")
        return choices


class QuizCreate(BaseModel):
    title: str
    passing_score: int = 70
    questions: list[QuestionCreate]

    @field_validator("questions")
    @classmethod
    def must_have_questions(cls, questions: list[QuestionCreate]) -> list[QuestionCreate]:
        if not questions:
            raise ValueError("A quiz needs at least one question")
        return questions


class ChoiceOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    text: str


class ChoiceOutWithAnswer(ChoiceOut):
    is_correct: bool


class QuestionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    text: str
    order_index: int
    choices: list[ChoiceOut]


class QuizOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    lesson_id: int
    title: str
    passing_score: int
    questions: list[QuestionOut]


class QuestionOutWithAnswers(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    text: str
    order_index: int
    choices: list[ChoiceOutWithAnswer]


class QuizOutWithAnswers(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    lesson_id: int
    title: str
    passing_score: int
    questions: list[QuestionOutWithAnswers]


class AnswerSubmission(BaseModel):
    question_id: int
    choice_id: int


class QuizSubmission(BaseModel):
    answers: list[AnswerSubmission]


class AnswerResult(BaseModel):
    question_id: int
    selected_choice_id: int | None
    correct_choice_id: int
    is_correct: bool


class QuizResult(BaseModel):
    attempt_id: int
    score: int
    total: int
    percentage: float
    passed: bool
    submitted_at: datetime
    answers: list[AnswerResult]


class QuizAttemptOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    quiz_id: int
    score: int
    total: int
    passed: bool
    submitted_at: datetime
